hw/ip/otbn/dv/rig/rig/gens/edge_load_store.py

```diff
@@ -0,0 +1,234 @@
+                self.bnsid.operands[0].op_type.reg_type == 'gpr' and
+                not self.bnsid.operands[0].op_type.is_dest() and
+                isinstance(self.bnsid.operands[1].op_type, RegOperandType) and
+                self.bnsid.operands[1].op_type.reg_type == 'gpr' and
+                not self.bnsid.operands[1].op_type.is_dest() and
+                isinstance(self.bnsid.operands[2].op_type, ImmOperandType) and
+                self.bnsid.operands[2].op_type.signed and
+                isinstance(self.bnsid.operands[3].op_type, OptionOperandType) and
+                isinstance(self.bnsid.operands[4].op_type, OptionOperandType)):
+            raise RuntimeError('BN.SID instruction from instructions file is'
+                               'not the shape expected by MisalignedBadLoadStore'
+                               'generator.')
+
+        for insn in insns_file.insns:
+            if (insn.mnemonic in ['lw', 'sw', 'bn.lid', 'bn.sid']):
+                weight = cfg.insn_weights.get(insn.mnemonic)
+                if weight > 0:
+                    self.weights.append(weight)
+                    self.insns.append(insn)
+
+        # Check that at least one instruction has a positive weight
+        assert len(self.insns) == len(self.weights)
+        if not self.weights:
+            raise ValueError('Config at {} defines a zero weight '
+                             'for all instructions.'
+                             .format(cfg.path))
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 1:
+            return None
+
+        # Determine the offset range for XW
+        offset_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for XW (-2048, 2047)
+        max_offset = offset_rng[1]
+        min_offset = offset_rng[0]
+
+        # Determine the offset range for BN.XID
+        bn_offset_rng = self.bn_imm_op_type.get_op_val_range(model.pc)
+        assert bn_offset_rng is not None
+
+        # Max/Min Offsets for BN.XID (-512 * 32, 511 * 32)
+        bn_max_offset = bn_offset_rng[1]
+        bn_min_offset = bn_offset_rng[0]
+
+        # Pick a random instruction from the weighted list
+        insn_idx = random.choices(range(len(self.insns)),
+                                  weights=self.weights)[0]
+
+        imm_val = 0
+        bn_imm_val = 0
+        base = []
+        wdr_base = []
+
+        known_regs = model.regs_with_known_vals('gpr')
+        # Pick a near the top known value for hitting edges
+        for reg_idx, reg_val in known_regs:
+            if (4096 + max_offset >= reg_val >= 4096 + min_offset):
```

Comment:
As above, could you grab the size of DMEM from `program`?

Also, I think this is trying to pick a register value which can be used to get the address `DMEM_SIZE - 4`. So I think you actually want `min_offset <= DMEM_SIZE - reg_val <= max_offset`?

(And, total nit, I'd generally go for `<=` and `<` instead of `>=` and `>`, because then the little numbers appear on the left. But that might just be my maths background!)